modal/_functions.py

```diff
@@ -501,18 +501,43 @@ def from_args(
             if include_source_mode != IncludeSourceMode.INCLUDE_NOTHING:
                 entrypoint_mounts = info.get_entrypoint_mount()
             else:
-                entrypoint_mounts = []
+                entrypoint_mounts = {}
 
             all_mounts = [
                 _get_client_mount(),
                 *explicit_mounts,
-                *entrypoint_mounts,
+                *entrypoint_mounts.values(),
             ]
 
             if include_source_mode is IncludeSourceMode.INCLUDE_FIRST_PARTY:
-                # TODO(elias): if using INCLUDE_FIRST_PARTY *and* mounts are added that haven't already been
-                #  added to the image via add_local_python_source
-                all_mounts += get_auto_mounts()
+                auto_mounts = get_sys_modules_mounts()
+                # don't need to add entrypoint modules to automounts:
+                for entrypoint_module in entrypoint_mounts:
+                    auto_mounts.pop(entrypoint_module, None)
+
+                warn_missing_modules = set(auto_mounts.keys()) - image._added_python_source_set
+
+                if warn_missing_modules:
+                    python_stringified_modules = ", ".join(f'"{mod}"' for mod in sorted(warn_missing_modules))
+                    deprecation_warning(
+                        (2024, 1, 31),
+                        (
+                            "Automatic adding of local python source will be deprecated in the future.\n"
+                            f"Make sure you have explicitly added the source for the following modules to the Image "
+                            f"used by `{info.function_name}`:\n"
+                        )
+                        + ", ".join(sorted(warn_missing_modules))
+                        + "\n\n"
+                        + (
+                            "e.g.:\n"
+                            f"image = Image.debian_slim().add_local_python_source"
+                            f"({python_stringified_modules})\n"
+                            f"@app.function(image=image)\n"
+                            f"..."
+                        ),
```

Comment:
Can we add a line like "For more information, see https://modal.com/docs/guide/modal-1-0-migration"? When users start seeing all the new deprecation warnings I think it will be helpful to call their attention to the idea that this is "one last hurrah" to manage their annoyance :D.

We can also add a more verbose explanation of what's going on and the various options for ameliorating it there. I can work on that.